In [2]:
#!pip install plotly

In [3]:
import pandas as pd
import numpy as np
pd.set_option("display.max_columns", None)
from datetime import datetime

import plotly.graph_objects as go
from plotly.subplots import make_subplots

import seaborn as sns
import matplotlib.pyplot as plt
plt.rcParams.update({'font.size': 14})

from statsmodels.graphics.tsaplots import plot_acf
from statsmodels.graphics.tsaplots import plot_pacf

In [4]:
## Preprocessing Libraries
from sklearn.model_selection import train_test_split
from sklearn.impute import KNNImputer
from sklearn.feature_selection import VarianceThreshold
from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler


# READING DATA 

In [5]:
df1 = pd.read_csv('./asset_details.csv')
df2 = pd.read_csv('./example_sample_submission.csv')
df3 = pd.read_csv('./example_test.csv')
df4 = pd.read_csv('./supplemental_train.csv')
df5 =  pd.read_csv('./train.csv')


In [6]:
df5

timestamp  Asset_ID  Count          Open          High  \
0         1514764860         2   40.0   2376.580000   2399.500000   
1         1514764860         0    5.0      8.530000      8.530000   
2         1514764860         1  229.0  13835.194000  14013.800000   
3         1514764860         5   32.0      7.659600      7.659600   
4         1514764860         7    5.0     25.920000     25.920000   
...              ...       ...    ...           ...           ...   
24236801  1632182400         9  775.0    157.181571    157.250000   
24236802  1632182400        10   34.0   2437.065067   2438.000000   
24236803  1632182400        13  380.0      0.091390      0.091527   
24236804  1632182400        12  177.0      0.282168      0.282438   
24236805  1632182400        11   48.0    232.695000    232.800000   

                   Low         Close        Volume          VWAP    Target  
0          2357.140000   2374.590000  1.923301e+01   2373.116392 -0.004218  
1             8.530000      8.530000  7.838000e+01      8.530000 -0.014399  
2         13666.110000  13850.176000  3.155006e+01  13827.062093 -0.014643  
3             7.656700      7.657600  6.626713e+03      7.657713 -0.013922  
4            25.874000     25.877000  1.210873e+02     25.891363 -0.008264  
...                ...           ...           ...           ...       ...  
24236801    156.700000    156.943857  4.663725e+03    156.994319       NaN  
24236802   2430.226900   2432.907467  3.975460e+00   2434.818747       NaN  
24236803      0.091260      0.091349  2.193732e+06      0.091388       NaN  
24236804      0.281842      0.282051  1.828508e+05      0.282134       NaN  
24236805    232.240000    232.275000  1.035123e+02    232.569697       NaN  

[24236806 rows x 10 columns]

In [7]:
import collections 

In [8]:
dict = {}

In [9]:
df5

timestamp  Asset_ID  Count          Open          High  \
0         1514764860         2   40.0   2376.580000   2399.500000   
1         1514764860         0    5.0      8.530000      8.530000   
2         1514764860         1  229.0  13835.194000  14013.800000   
3         1514764860         5   32.0      7.659600      7.659600   
4         1514764860         7    5.0     25.920000     25.920000   
...              ...       ...    ...           ...           ...   
24236801  1632182400         9  775.0    157.181571    157.250000   
24236802  1632182400        10   34.0   2437.065067   2438.000000   
24236803  1632182400        13  380.0      0.091390      0.091527   
24236804  1632182400        12  177.0      0.282168      0.282438   
24236805  1632182400        11   48.0    232.695000    232.800000   

                   Low         Close        Volume          VWAP    Target  
0          2357.140000   2374.590000  1.923301e+01   2373.116392 -0.004218  
1             8.530000      8.530000  7.838000e+01      8.530000 -0.014399  
2         13666.110000  13850.176000  3.155006e+01  13827.062093 -0.014643  
3             7.656700      7.657600  6.626713e+03      7.657713 -0.013922  
4            25.874000     25.877000  1.210873e+02     25.891363 -0.008264  
...                ...           ...           ...           ...       ...  
24236801    156.700000    156.943857  4.663725e+03    156.994319       NaN  
24236802   2430.226900   2432.907467  3.975460e+00   2434.818747       NaN  
24236803      0.091260      0.091349  2.193732e+06      0.091388       NaN  
24236804      0.281842      0.282051  1.828508e+05      0.282134       NaN  
24236805    232.240000    232.275000  1.035123e+02    232.569697       NaN  

[24236806 rows x 10 columns]

In [10]:
train = df5.copy()

In [11]:
train.describe()

timestamp      Asset_ID         Count          Open          High  \
count  2.423681e+07  2.423681e+07  2.423681e+07  2.423681e+07  2.423681e+07   
mean   1.577120e+09  6.292544e+00  2.864593e+02  1.432640e+03  1.436350e+03   
std    3.323350e+07  4.091861e+00  8.673982e+02  6.029605e+03  6.039482e+03   
min    1.514765e+09  0.000000e+00  1.000000e+00  1.170400e-03  1.195000e-03   
25%    1.549011e+09  3.000000e+00  1.900000e+01  2.676500e-01  2.681600e-01   
50%    1.578372e+09  6.000000e+00  6.400000e+01  1.428860e+01  1.431250e+01   
75%    1.606198e+09  9.000000e+00  2.210000e+02  2.288743e+02  2.293000e+02   
max    1.632182e+09  1.300000e+01  1.650160e+05  6.480594e+04  6.490000e+04   

                Low         Close        Volume          VWAP        Target  
count  2.423681e+07  2.423681e+07  2.423681e+07  2.423680e+07  2.348647e+07  
mean   1.429568e+03  1.432640e+03  2.868530e+05           NaN  7.121752e-06  
std    6.020261e+03  6.029611e+03  2.433935e+06           NaN  5.679042e-03  
min    2.000000e-04  1.171400e-03 -3.662812e-01          -inf -5.093509e-01  
25%    2.669000e-01  2.676484e-01  1.410725e+02  2.676368e-01 -1.694354e-03  
50%    1.426300e+01  1.428920e+01  1.295415e+03  1.428769e+01 -4.289844e-05  
75%    2.284200e+02  2.288729e+02  2.729764e+04  2.288728e+02  1.601520e-03  
max    6.467053e+04  6.480854e+04  7.597554e+08           inf  9.641699e-01

In [19]:
train['Asset_ID'].nunique()

14

there are 14 crypto currencies in the whole data set so Asset ID = 14 

## RESAMPELLING

For the following EDA, we will **resample the minute-wise crypto data to daily samples**. This reduces the amount of samples from 24,236,806 to 1,360. 

In [12]:
train['timestamp'] = train['timestamp'].astype('datetime64[s]')

In [16]:
#initiaizing an empty data frame 

train_daily = pd.DataFrame() 

In [15]:
train_daily.shape

(0, 0)

In [ ]:
for asset_id in asset_details.Asset_ID:
    train_single = train[train.Asset_ID == asset_id].copy()

    train_single_new = train_single[['timestamp','Count']].resample('D', on='timestamp').sum()
    train_single_new['Open'] = train_single[['timestamp','Open']].resample('D', on='timestamp').first()['Open']
    train_single_new['High'] = train_single[['timestamp','High']].resample('D', on='timestamp').max()['High']
    train_single_new['Low'] = train_single[['timestamp','Low']].resample('D', on='timestamp').min()['Low']
    train_single_new['Close'] = train_single[['timestamp','Close']].resample('D', on='timestamp').last()['Close']
    train_single_new['Volume'] = train_single[['timestamp','Volume']].resample('D', on='timestamp').sum()['Volume']
    # train_single_new['VWAP']
    #train_single_new['Target'] = train_single[['timestamp','Target']].resample('D', on='timestamp').mean()['Target']
    train_single_new['Asset_ID'] = asset_id

    train_daily = train_daily.append(train_single_new.reset_index(drop=False))
train_daily = train_daily.sort_values(by = ['timestamp', 'Asset_ID']).reset_index(drop=True)

train_daily = train_daily.pivot(index='timestamp', columns='Asset_ID')[['Count', 'Open', 'High', 'Low', 'Close', 'Volume']]
train_daily = train_daily.reset_index(drop=False)

display(train_daily.head(10))